### DISCUSSION TITLE
Zenith: enable built-in SDKs to be usable without hard dependency on the internet

### DISCUSSION THREAD

**Discussion by sipsma** [dagger/dagger#12393] [maintainers]
Some users hit roadblocks when the dagger engine has hard dependencies on pulling e.g. container images from dockerhub or other resources from the internet at runtime. This is typically due to stringent corporate firewall rules that require exceptions be made for any unapproved network calls to the internet.

Our builtin module SDKs (Go + Python currently, nodejs soon), currently have varying degrees of dependencies on base images + library dependencies that result in network calls to the internet (at least on initial use, should be cached locally after unless pruned).

https://github.com/dagger/dagger/pull/5900 did some of the initial leg work required to support removal of internet dependencies, but it's not complete yet.

My most recent idea on an approach (not fully fleshed out, subject to change + suggestions):

> The approach I have in mind is to use another optional Function on SDK Modules, Base, that returns a Container with everything needed to run the Module w/out internet deps. Then, we can add a mechanism to our CI for calling those, exporting to a local cache and including that cache in the engine image.

**Comment by d3rp3tt3** [comment]
Discussed in this Discord thread: https://discord.com/channels/707636530424053791/1218318319409430529

**Comment by Orrimp** [comment]
Hello, we also need Air Gapped Environment Running for Dagger. When running it in our Jenkins infrastructure the error appears: 

```
[95m1: [0m[2m07:29:19[0m [93mWRN[0m failed to resolve image; falling back to leftover engine [2merror=[0m"Get \"[https://registry.dagger.io/v2/\](https://registry.dagger.io/v2//)": dial tcp XX.X.X.XX:443: connect: connection refused"
09:29:19  [95m1: [0m[2m07:29:19[0m [93mWRN[0m failed to list containers [2merror=[0m"exit status 1"
```
I read somewhere that fer `docker login` the repository for images can be changed, but it did not work. 

Defining the `From()` function with hole artefact URL to private repo also failed. Also using `WithRegistryAuth` failed. 
Maybe I did something wrong, I will try further and post a solution here
